File: problems-extend/3701-3800/3779_minimum-number-of-operations-to-have-distinct-elements/solution.go
// Pointer + counts: counts tracks the remaining suffix, duplicated how
// many distinct values it still holds twice or more. While the suffix has
// a duplicate, one operation advances the pointer by three and refreshes
// only those three values (the last, possibly shorter, operation removes
// whatever is left).
func minOperations(nums []int) int {
	counts := make(map[int]int)
	for _, v := range nums {
		counts[v]++
	}
	duplicated := 0
	for _, c := range counts {
		if c >= 2 {
			duplicated++
		}
	}
	i := 0
	ops := 0
	n := len(nums)
	for i < n && duplicated > 0 {
		end := i + 3
		if end > n {
			end = n
		}
		for j := i; j < end; j++ {
			counts[nums[j]]--
			if counts[nums[j]] == 1 {
				duplicated--
			}
		}
		i += 3
		ops++
	}
	return ops
}

File: problems-extend/3701-3800/3779_minimum-number-of-operations-to-have-distinct-elements/solution.py
from typing import List


class Solution:
    def minOperations(self, nums: List[int]) -> int:
        # Pointer + counts: counts tracks the remaining suffix, duplicated
        # how many distinct values it still holds twice or more. While the
        # suffix has a duplicate, one operation advances the pointer by
        # three and refreshes only those three values (the last, possibly
        # shorter, operation removes whatever is left).
        counts = {}
        for v in nums:
            counts[v] = counts.get(v, 0) + 1
        duplicated = sum(1 for c in counts.values() if c >= 2)
        i = 0
        ops = 0
        n = len(nums)
        while i < n and duplicated > 0:
            for j in range(i, min(i + 3, n)):
                counts[nums[j]] -= 1
                if counts[nums[j]] == 1:
                    duplicated -= 1
            i += 3
            ops += 1
        return ops
